kotest-assertions-arrow-fx-coroutines/src/commonMain/kotlin/io/kotest/assertions/arrow/fx/coroutines/ResourceExtension.kt

```diff
@@ -0,0 +1,170 @@
+ *     { Resource(name) },
+ *     { conn, ex -> conn.release(ex) }
+ *   )
+ *
+ * class ResourceSpec : StringSpec({
+ *   val conn: TestResource<Connection> =
+ *     install(ResourceExtension(connection("DB")))
+ *
+ *   "invoke" {
+ *     conn().compute() shouldBe "Result of DB"
+ *   }
+ *
+ *   "get" {
+ *     conn.get().compute() shouldBe "Result of DB"
+ *   }
+ *
+ *   afterSpec {
+ *     shouldThrow<TestResource.AlreadyClosedException> {
+ *       conn.get()
+ *     }
+ *   }
+ * })
+ * ```
+ *
+ * `conn` cannot be used during or after `afterSpec` since at that point the resource is already released,
+ * and it will result in [IllegalStateException]
+ *
+ * @param lifecycleMode allows installing the [Resource] for other scopes besides the default [LifecycleMode.Spec].
+ */
+public class ResourceExtension<A>(
+  public val resource: Resource<A>, public val lifecycleMode: LifecycleMode = LifecycleMode.Spec,
+) : LazyMountableExtension<A, A>, TestListener, AfterSpecListener {
+  
+  private var underlying: ResourceLazyMaterialized<A>? = null
+  
+  override fun mount(configure: A.() -> Unit): LazyMaterialized<A> =
+    ResourceLazyMaterialized(resource, configure).also {
+      underlying = it
+    }
+  
+  override suspend fun beforeSpec(spec: Spec) {
+    super.beforeSpec(spec)
+    if (lifecycleMode == LifecycleMode.Spec) {
+      underlying?.init()
+    }
+  }
+  
+  override suspend fun afterSpec(spec: Spec) {
+    if (lifecycleMode == LifecycleMode.Spec) {
+      underlying?.release()
+    }
+    underlying?.close()
+  }
+  
+  override suspend fun beforeAny(testCase: TestCase) {
+    val every = lifecycleMode == LifecycleMode.EveryTest
+    val root = lifecycleMode == LifecycleMode.Root && testCase.isRootTest()
+    val leaf = lifecycleMode == LifecycleMode.Leaf && testCase.type == TestType.Test
+    if (every || root || leaf) {
+      underlying?.init()
+    }
+  }
+  
+  override suspend fun afterAny(testCase: TestCase, result: TestResult) {
+    val every = lifecycleMode == LifecycleMode.EveryTest
+    val root = lifecycleMode == LifecycleMode.Root && testCase.isRootTest()
+    val leaf = lifecycleMode == LifecycleMode.Leaf && testCase.type == TestType.Test
+    if (every || root || leaf) {
+      underlying?.release()
+    }
+  }
+}
+
+internal class ResourceLazyMaterialized<A>(
+  private val resource: Resource<A>,
+  private val configure: A.() -> Unit = {},
+) : LazyMaterialized<A> {
+  
+  sealed interface State<out A> {
+    object Empty : State<Nothing>
+    object Closed : State<Nothing>
+    data class Loading<A>(
+      val acquiring: CompletableDeferred<A> = CompletableDeferred(),
+      val finalizers: CompletableDeferred<suspend (ExitCase) -> Unit> = CompletableDeferred(),
+    ) : State<A>
+    
+    data class Done<A>(val value: A, val finalizers: suspend (ExitCase) -> Unit) : State<A>
+  }
+  
+  private val state = AtomicReference<State<A>>(State.Empty)
+  
+  override suspend fun get(): A = init()
```

Comment:
```suggestion
  override suspend fun acquire(): A = init()
```

Just throwing the idea out there.. I believe some tests referred to "acquiring" the resource

Reply:
This is an `override` from `LazyMountableExtension` so we cannot change the name.